tsdb/query_executor.go

```diff
@@ -547,63 +550,63 @@ func (q *QueryExecutor) filterShowSeriesResult(limit, offset int, rows influxql.
 	return filteredSeries
 }
 
-func (q *QueryExecutor) executeShowMeasurementsStatement(stmt *influxql.ShowMeasurementsStatement, database string) *influxql.Result {
-	// Find the database.
-	db := q.Store.DatabaseIndex(database)
-	if db == nil {
-		return &influxql.Result{}
+// PlanShowMeasurements creates an execution plan for the given SelectStatement and returns an Executor.
+func (q *QueryExecutor) PlanShowMeasurements(stmt *influxql.ShowMeasurementsStatement, database string, chunkSize int) (Executor, error) {
+	// Get the database info.
+	fmt.Printf("database = %s\n", database)
+	di, err := q.MetaStore.Database(database)
+	if err != nil {
+		return nil, err
+	} else if di == nil {
+		return nil, ErrDatabaseNotFound(database)
 	}
 
-	var measurements Measurements
+	// Get info for all shards in the database.
+	shards := di.ShardInfos()
 
-	// If a WHERE clause was specified, filter the measurements.
-	if stmt.Condition != nil {
-		var err error
-		measurements, err = db.measurementsByExpr(stmt.Condition)
+	// Build the Mappers, one per shard.
+	mappers := []Mapper{}
+	for _, sh := range shards {
+		m, err := q.ShardMapper.CreateMapper(sh, stmt, chunkSize)
 		if err != nil {
-			return &influxql.Result{Err: err}
+			return nil, err
 		}
-	} else {
-		// Otherwise, get all measurements from the database.
-		measurements = db.Measurements()
+		if m == nil {
+			// No data for this shard, skip it.
+			continue
+		}
+		mappers = append(mappers, m)
 	}
-	sort.Sort(measurements)
 
-	offset := stmt.Offset
-	limit := stmt.Limit
-
-	// If OFFSET is past the end of the array, return empty results.
-	if offset > len(measurements)-1 {
-		return &influxql.Result{}
-	}
+	executor := NewShowMeasurementsExecutor(stmt, mappers, chunkSize)
```

Comment:
For now, yes.  Maybe once it's all written it will be obvious where some duplicate code can be removed.  I started down the path of trying to avoid this and it became messy.  Also, doing it this way is less risky because no big changes have to be made to `SelectMapper` (used to be `LocalMapper`) or that part of the query engine.